pkg/util/acrtoken/acrtoken.go

```diff
@@ -0,0 +1,113 @@
+}
+
+func NewManager(env env.Interface, fpAuthorizer autorest.Authorizer) Manager {
+	return &manager{
+		env: env,
+
+		tokens:     containerregistry.NewTokensClient(env.SubscriptionID(), fpAuthorizer),
+		registries: containerregistry.NewRegistriesClient(env.SubscriptionID(), fpAuthorizer),
+
+		fullRegistryName: env.ACRName() + ".azurecr.io",
+		scopeMap:         fmt.Sprintf("/subscriptions/%s/resourceGroups/%s/providers/Microsoft.ContainerRegistry/registries/%s/scopeMaps/_repositories_pull", env.SubscriptionID(), env.ACRResourceGroup(), env.ACRName()),
+	}
+}
+
+func (m *manager) tokenResourceID(name string) string {
+	return fmt.Sprintf("/subscriptions/%s/resourceGroups/%s/providers/Microsoft.ContainerRegistry/registries/%s/tokens/%s", m.env.SubscriptionID(), m.env.ACRResourceGroup(), m.env.ACRName(), name)
+}
+
+func (m *manager) generateTokenName(clusterName string) (string, error) {
+	rs, err := stringutils.RandomString("abcdefghijklmnopqrstuvwxyz", 8)
+	if err != nil {
+		return "", err
+	}
+	return clusterName[:compareutils.Min(len(clusterName), 8)] + "-" + rs, nil
+}
+
+// Create create a token on our registry
+// see https://docs.microsoft.com/en-us/azure/container-registry/container-registry-repository-scoped-permissions
+func (m *manager) Create(ctx context.Context, oc *api.OpenShiftCluster) error {
+	for _, pss := range oc.Properties.RegistryProfiles {
+		if pss.RegistryName == m.fullRegistryName {
+			// already created
+			return nil
+		}
+	}
+
+	acrTokenName, err := m.generateTokenName(oc.Name)
+	if err != nil {
+		return err
+	}
+	err = m.tokens.CreateAndWait(ctx, m.env.ACRResourceGroup(), m.env.ACRName(), acrTokenName, azcontainerregistry.Token{
+		TokenProperties: &azcontainerregistry.TokenProperties{
+			ScopeMapID: &m.scopeMap,
+			Status:     azcontainerregistry.TokenStatusEnabled,
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	generateCredentialsParameters := azcontainerregistry.GenerateCredentialsParameters{
+		TokenID: to.StringPtr(m.tokenResourceID(acrTokenName)),
+		Name:    azcontainerregistry.TokenPasswordNamePassword1,
+	}
+	creds, err := m.registries.GenerateCredentials(ctx, m.env.ACRResourceGroup(), m.env.ACRName(), generateCredentialsParameters)
+	if err != nil {
+		return err
+	}
+	if creds.Passwords == nil || len(*creds.Passwords) < 1 || (*creds.Passwords)[0].Value == nil {
+		return fmt.Errorf("generateCredentials returned empty passwords")
+	}
+	acrPassword := *(*creds.Passwords)[0].Value
+	oc.Properties.RegistryProfiles = append(oc.Properties.RegistryProfiles, api.RegistryProfile{
+		RegistryName: m.fullRegistryName,
+		Username:     acrTokenName,
+		Password:     api.SecureString(acrPassword),
+	})
+	return nil
+}
+
+func (m *manager) Delete(ctx context.Context, oc *api.OpenShiftCluster) error {
+	for _, pss := range oc.Properties.RegistryProfiles {
+		if pss.RegistryName == m.fullRegistryName {
+			err := m.tokens.DeleteAndWait(ctx, m.env.ACRResourceGroup(), m.env.ACRName(), pss.Username)
```

Comment:
please ensure this doesn't return an error if the token doesn't exist at the point where DeleteAndWait is called - e.g. if Delete is called more than once.